mmdet/models/dense_heads/dino_head.py

```diff
@@ -0,0 +1,307 @@
+            hidden_states (Tensor): Hidden states output from each decoder
+                layer, has shape (num_decoder_layers, num_query, bs, dim).
+            references (list[Tensor]): List of the reference from the decoder.
+                The first reference is the `init_reference` (initial) and the
+                other num_decoder_layers(6) references are `inter_references`
+                (intermediate). The `init_reference` has shape (bs, num_query,
+                4) when `as_two_stage` of the detector is `True`, otherwise
+                (bs, num_query, 2). Each `inter_reference` has shape
+                (bs, num_query, 4) when `with_box_refine` of the detector is
+                `True`, otherwise (bs, num_query, 2).
+            enc_outputs_class (Tensor): The score of each point on encode
+                feature map, has shape (bs, num_feat, cls_out_channels).
+                Only when `as_two_stage` is `True` it would be returned,
+                otherwise `None` would be returned.
+            enc_outputs_coord (Tensor): The proposal generate from the
+                encode feature map, has shape (bs, num_feat, 4). Only when
+                `as_two_stage` is `True` it would be returned, otherwise
+                `None` would be returned.
+            batch_data_samples (list[:obj:`DetDataSample`]): The Data
+                Samples. It usually includes information such as
+                `gt_instance`, `gt_panoptic_seg` and `gt_sem_seg`.
+            dn_meta (Dict): TODO: write doc
+
+        Returns:
+            dict: A dictionary of loss components.
+        """
+        batch_gt_instances = []
+        batch_img_metas = []
+        for data_sample in batch_data_samples:
+            batch_img_metas.append(data_sample.metainfo)
+            batch_gt_instances.append(data_sample.gt_instances)
+
+        outs = self(hidden_states, references)
+        loss_inputs = outs + (enc_outputs_class, enc_outputs_coord,
+                              batch_gt_instances, batch_img_metas, dn_meta)
+        losses = self.loss_by_feat(*loss_inputs)
+        return losses
+
+    def loss_by_feat(
+        self,
+        all_layers_cls_scores: Tensor,
+        all_layers_bbox_preds: Tensor,
+        enc_cls_scores: Tensor,
+        enc_bbox_preds: Tensor,
+        batch_gt_instances: InstanceList,
+        batch_img_metas: List[dict],
+        dn_meta: Dict,  # TODO: Dict[what]
+        batch_gt_instances_ignore: OptInstanceList = None
+    ) -> Dict[str, Tensor]:
+        """Loss function.
+
+        Args:
+            all_layers_cls_scores (Tensor): Classification scores of all
+                decoder layers, has shape (num_decoder_layers, bs, num_query,
+                cls_out_channels).
+            all_layers_bbox_preds (Tensor): Regression outputs of all decode
+                layers. Each is a 4D-tensor with normalized coordinate format
+                (cx, cy, w, h) and has shape (num_decoder_layers, bs,
```

Comment:
has shape (num_decoder_layers, bs,num_qeury, 4) with the last dimension arranged as (cx, cy, w, h)

Reply:
OK, I have added the coordinate format to all the related comments.